FIAT/alfeld_sorokina.py

```diff
@@ -0,0 +1,83 @@
+# Copyright (C) 2024 Pablo D. Brubeck
+#
+# This file is part of FIAT (https://www.fenicsproject.org)
+#
+# SPDX-License-Identifier:    LGPL-3.0-or-later
+#
+# Written by Pablo D. Brubeck (brubeck@protonmail.com), 2024
+
+from FIAT import finite_element, dual_set, polynomial_set
+from FIAT.functional import ComponentPointEvaluation, PointDivergence
+from FIAT.quadrature_schemes import create_quadrature
+from FIAT.macro import CkPolynomialSet, AlfeldSplit
+
+import numpy
+
+
+def AlfeldSorokinaSpace(ref_el, degree):
+    """Return a vector-valued C^0 PolynomialSet on an Alfeld split whose
+    divergence is also C^0.  This works for on any simplex and for all
+    polynomial degrees."""
+    ref_complex = AlfeldSplit(ref_el)
+    sd = ref_complex.get_spatial_dimension()
+    C0 = CkPolynomialSet(ref_complex, degree, order=0, shape=(sd,), variant="bubble")
+    expansion_set = C0.get_expansion_set()
+    num_members = C0.get_num_members()
+    coeffs = C0.get_coeffs()
+
+    facet_el = ref_complex.construct_subelement(sd-1)
+    phi = polynomial_set.ONPolynomialSet(facet_el, 0 if sd == 1 else degree-1)
+    Q = create_quadrature(facet_el, 2 * phi.degree)
+    qpts, qwts = Q.get_points(), Q.get_weights()
+    phi_at_qpts = phi.tabulate(qpts)[(0,) * (sd-1)]
+    weights = numpy.multiply(phi_at_qpts, qwts)
+
+    rows = []
+    for facet in ref_complex.get_interior_facets(sd-1):
+        n = ref_complex.compute_normal(facet)
+        jumps = expansion_set.tabulate_normal_jumps(degree, qpts, facet, order=1)
+        div_jump = n[:, None, None] * jumps[1][None, ...]
+        r = numpy.tensordot(div_jump, weights, axes=(-1, -1))
+        rows.append(r.reshape(num_members, -1).T)
+
+    if len(rows) > 0:
+        dual_mat = numpy.vstack(rows)
+        _, sig, vt = numpy.linalg.svd(dual_mat, full_matrices=True)
+        tol = sig[0] * 1E-10
+        num_sv = len([s for s in sig if abs(s) > tol])
+        coeffs = numpy.tensordot(vt[num_sv:], coeffs, axes=(-1, 0))
+    return polynomial_set.PolynomialSet(ref_complex, degree, degree, expansion_set, coeffs)
+
+
+class AlfeldSorokinaDualSet(dual_set.DualSet):
+    def __init__(self, ref_el, degree):
+        if degree != 2:
```

Comment:
You have the space for any degree but the dual space bails out if `degree != 2`.

Reply:
And should this be a `ValueError` or `NotImplemented` error?

Reply:
In theory, it should be posible to define the element for high order for the same reasons high-order HCT is a thing. Nevertheless we might need to impose some supersmoothness at the barycenter. I think this could be part of a future PR, so I am going to raise a `NotImplementedError`.